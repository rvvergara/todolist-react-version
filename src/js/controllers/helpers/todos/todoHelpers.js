import {
  getDataFromLocalStorage,
  setDataIntoLocalStorage,
} from '../common/storage';

const parseDate = (dateStr) => {
  const currentDate = new Date(dateStr);
  const monthPrefix = currentDate.getMonth() < 10 ? '0' : '';
  const dayPrefix = currentDate.getDate() < 10 ? '0' : '';
  return `${currentDate.getFullYear()}-${monthPrefix}${currentDate.getMonth() + 1}-${dayPrefix}${currentDate.getDate()}`;
};

export const createTodosArray = () => {
  if (getDataFromLocalStorage('todosArray') === null) {
    const todosArray = [];
    setDataIntoLocalStorage('todosArray', todosArray);
  }
};

export const getTodoDataFromForm = (name) => {
  const inputs = document.getElementsByClassName('todo-form')[0].elements;
  const title = inputs[0].value;
  const description = inputs[1].value;
  const dueDate = parseDate(inputs[2].value);
  const priority = document.getElementsByTagName('select')[0].value;
  const notes = inputs[4].value;
  return [
    title,
    description,
    dueDate,
    priority,
    notes,
    name,
  ];
};

export const updateProjectsArray = (project) => {
  const projectsArr = getDataFromLocalStorage('projectsArray');
  const projectIndex = projectsArr.findIndex(x => x.id === project.id);
  projectsArr[projectIndex] = project;
  setDataIntoLocalStorage('projectsArray', projectsArr);
  return projectsArr;
};

export const pushTodoToProject = (todo) => {
  const projectsArray = getDataFromLocalStorage('projectsArray');
  const todosArray = getDataFromLocalStorage('todosArray');
  const todoProjectInArray = projectsArray.find(project => project.id === todo.projectID);
  const parentProjectInArray = getDataFromLocalStorage(todoProjectInArray.name);
  const parentProjectStandAlone = getDataFromLocalStorage(parentProjectInArray.name);
  parentProjectInArray.todos.push(todo);
  todosArray.push(todo);
  parentProjectStandAlone.todos.push(todo);
  setDataIntoLocalStorage(todoProjectInArray.name, parentProjectStandAlone);
  updateProjectsArray(parentProjectInArray);
  setDataIntoLocalStorage('todosArray', todosArray);
};

export const updateTodoInProject = (todo, project, index) => {
  const {
    todos,
  } = project;
  todos[index] = todo;
  const todosArray = getDataFromLocalStorage('todosArray');
  const todosArrayIndex = todosArray.findIndex(t => t.id === todo.id);
  todosArray[todosArrayIndex] = todo;
  setDataIntoLocalStorage(project.name, project);
  setDataIntoLocalStorage('todosArray', todosArray);
  updateProjectsArray(project);
  return project;
};
