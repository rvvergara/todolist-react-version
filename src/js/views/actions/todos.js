import uuid from 'uuid';

const addTodo = (todo, projectName) => ({
  type: 'ADD_TODO',
  todo: {
    ...todo,
    projectName,
    id: uuid(),
  },
});

const updateTodo = (id, updates) => ({
  type: 'UPDATE_TODO',
  id,
  updates,
});

const deleteTodo = id => ({
  type: 'DELETE_TODO',
  id,
});

export {
  addTodo,
  updateTodo,
  deleteTodo,
};
