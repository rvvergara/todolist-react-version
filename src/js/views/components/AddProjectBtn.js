import React from 'react';
import PropTypes from 'prop-types';

const AddProjectBtn = (props) => {
  const {
    addProjectMode,
    clickAddProjectBtn,
  } = props;
  const displayClass = !addProjectMode ? 'btn btn-primary mt-1' : 'd-none';
  return (
    <button
      type="button"
      className={displayClass}
      onClick={clickAddProjectBtn}
    >
    Add New Project
    </button>
  );
};

AddProjectBtn.propTypes = {
  addProjectMode: PropTypes.bool.isRequired,
  clickAddProjectBtn: PropTypes.func.isRequired,
};

export default AddProjectBtn;
