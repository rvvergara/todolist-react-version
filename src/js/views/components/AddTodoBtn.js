import React from 'react';
import PropTypes from 'prop-types';

const AddTodoBtn = (props) => {
  const {
    handleTodoBtn,
  } = props;

  return (
    <button
      type="button"
      className="btn btn-primary btn-block"
      onClick={handleTodoBtn}
    >
    Add Todo
    </button>
  );
};

AddTodoBtn.propTypes = {
  handleTodoBtn: PropTypes.func.isRequired,
};

export default AddTodoBtn;
