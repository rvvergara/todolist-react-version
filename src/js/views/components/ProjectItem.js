import React from 'react';
import ProjectsForm from './ProjectsForm';
class ProjectItem extends React.Component {
  state = {
    todos: [],
    editProjectMode: false,
  }

  clickUpdateProjectBtn = () => {
    this.setState(() => ({
      editProjectMode: true,
    }));
  }

  submitProjectForm = (e) => {
    e.preventDefault();
    const id = Number(e.target.getAttribute('data-id'));
    this.props.updateProjectForm(id);
    this.setState(() => ({
      editProjectMode: false,
    }));
    e.target.reset();
  }

  render(){
    const {
      name,
      dataID,
      deleteProject,
      selectProject,
    } = this.props;
    const regular = (
      <div onClick={selectProject}>
        <span
          className="mx-1"
        >
          { name }
        </span>
        <button
          className="btn btn-sm btn-danger"
          id={`delete-proj-${dataID+1}`}
          onClick={deleteProject}
        >
          Delete
        </button>
        <button
          className="btn btn-sm btn-warning"
          onClick={this.clickUpdateProjectBtn}
        >
        Update
        </button>
      </div>
    );
    const editMode = (
      <ProjectsForm
      dataID={dataID+1}
      editProjectMode={this.state.editProjectMode}
      name={name}
      submitProjectForm={this.submitProjectForm}
      />
    )
    if (this.state.editProjectMode){
      return editMode
    }else{
      return regular
    }
  }
};

export default ProjectItem;