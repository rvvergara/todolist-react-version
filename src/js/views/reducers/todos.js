export default (state = [], action) => {
  switch (action.type) {
    case 'ADD_TODO':
      return [...state, action.todo];
    case 'UPDATE_TODO':
    {
      const todoIndex = state.findIndex(todo => todo.id === action.id);
      const updatedTodo = { ...state[todoIndex], ...action.updates };
      return [...state.slice(0, todoIndex), updatedTodo, ...state.slice(todoIndex + 1)];
    }
    case 'DELETE_TODO':
      return state.filter(todo => todo.id !== action.id);
    default:
      return state;
  }
};
