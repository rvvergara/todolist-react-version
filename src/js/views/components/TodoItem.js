import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';
import TodosForm from './TodosForm';
import todosController from '../../controllers/todosController';
import { updateTodoInProject } from '../../controllers/helpers/todos/todoHelpers';
class TodoItem extends React.Component {
  state = {
    todo: {...this.props.todo},
    done: false,
    editTodoMode: false,
    title: this.props.todo.title,
    description: this.props.todo.description,
    dueDate: this.props.todo.dueDate,
    priority: this.props.todo.priority,
    notes: this.props.todo.notes,
  }

  componentWillMount(){
    const todo = this.state.todo;
    this.setState(() => ({
      done: todo.done,
    }))
  }

  tickTodo = () => {
    this.setState((prevState) => ({
      done: !prevState.done,
    }));
    this.setState((prevState)=> {
      const newTodo = Object.assign({}, prevState.todo, {done: prevState.done});
      const project = JSON.parse(localStorage[newTodo.project]);
      const newTodoIndex = project.todos.findIndex(x => x.id === newTodo.id);
      updateTodoInProject(newTodo, project, newTodoIndex);
      return {
        todo: newTodo,
      }
    })
  }

  clickUpdateBtn = () => {
    this.setState(() => ({
      editTodoMode: true,
    }));
  }

  handleChange = (key, value) => {
    this.setState(() => ({
      [key]: value,
    }));
  }

  updateTodo = (e) => {
    e.preventDefault();
    const {
      todo
    } = this.state;
    
    const project = JSON.parse(localStorage[todo.project]);
    const todoID = todo.id;
    const projectName = project.name;
    const updatedTodo = todosController.update(project, todoID, projectName);
        
    this.setState(()=>({
      editTodoMode: false,
      title: updatedTodo.title,
      description: updatedTodo.description,
      dueDate: updatedTodo.dueDate,
      priority: updatedTodo.priority,
      notes: updatedTodo.notes,
    }));
  }

  render() {
    const {
      title,
      description,
      dueDate,
      priority,
      notes,
      done,
      todo,
    } = this.state;
    const doneClass = done ? 'strikeout' : '';
    const regular = (
      <tr
        className={doneClass}
        id={todo.id}
      >
        <td>{title}</td>
        <td>{description}</td>
        <td>{new Date(dueDate).toDateString()}</td>
        <td>{priority}</td>
        <td>{notes}</td>
        <td>
          <input
          type="checkbox"
          defaultValue={this.state.done}
          checked={this.state.done}
          onChange={this.tickTodo}
          />
        </td>
        <td>
          <button
            className="btn btn-sm btn-warning"
            disabled={this.state.done}
            onClick={this.clickUpdateBtn}
          >
            Update
          </button>
          <button
            className="btn btn-sm btn-danger"
            disabled={this.state.done}
            onClick={this.props.deleteTodo}
          >
            Delete
          </button>
        </td>
      </tr>
    );
    const editOutput = (
      <tr>
        <td colSpan="6">
          <TodosForm
          editTodoMode={this.state.editTodoMode}
          title={title}
          description={description}
          dueDate={dueDate}
          priority={priority}
          notes={notes}
          submitTodo={this.updateTodo}
          handleChange={this.handleChange}
          />
        </td>
      </tr>
      
    )
    return this.state.editTodoMode ? editOutput : regular; 
  }
}

export default TodoItem;