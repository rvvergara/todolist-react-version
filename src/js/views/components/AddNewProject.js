import React from 'react';
import { connect } from 'react-redux';
import AddProjectBtn from './AddProjectBtn';
import ProjectsForm from './ProjectsForm';
import {
  addProject,
} from '../actions/projects';
import {
  selectProject
} from '../actions/selectedProject';
import { addProjectModeSwitch, editProjectModeSwitch } from '../actions/projectForm';
import { addTodoModeSwitch, editTodoModeSwitch } from '../actions/todoForm';
import projectsController from '../../controllers/projectsController';

export class AddNewProject extends React.Component {
  state = {
    addProjectMode: false,
    error: '',
    projectName: '',
  };

  handleClickAddBtn = () => {
    this.props.addProjectModeSwitch();
    if(this.props.editProjectMode) this.props.editProjectModeSwitch();
    if(this.props.addTodoMode) this.props.addTodoModeSwitch();
    if(this.props.editTodoMode) this.props.editTodoModeSwitch();
  };

  handleChange = (key, val) => this.setState({
    [key]: val,
  });

  handleSubmitProjectForm = (e) => {
    e.preventDefault();
    const { addProject } = this.props;
    const name = this.state.projectName;
    const isValid = this.validateName(name);
    if(isValid && name){
      this.setState({
        projectName: '',
        error: '',
      });
      this.props.addProjectModeSwitch();
      const newProject = addProject({name});
      projectsController.create(name, newProject.project.id);
      this.props.selectProject(newProject.project);
      this.props.addNewProject(newProject.project);
    } else {
      this.setState({
        error: 'You have entered a duplicate or invalid name'
      });
    };
  };

  validateName = name => {
    const {projects} = this.props;
    return projects.findIndex(project => project.name === name) === -1; 
  };

  render() {
    return (
      <div>
        {!this.props.addProjectMode && <AddProjectBtn
          addProjectMode={this.state.addProjectMode}
          clickAddProjectBtn={this.handleClickAddBtn}
        />}
        { this.props.addProjectMode && <ProjectsForm
          addProjectMode={this.state.addProjectMode}
          submitProjectForm={this.handleSubmitProjectForm}
          handleChange={this.handleChange}
          error={this.state.error}
          name={this.state.projectName}
        />}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  projects: state.projects,
  addProjectMode: state.projectForm.addProjectMode,
  editProjectMode: state.projectForm.editProjectMode,
  addTodoMode: state.todoForm.addTodoMode,
  editTodoMode: state.todoForm.editTodoMode,
});

export default connect(mapStateToProps, { addProject, addProjectModeSwitch, editProjectModeSwitch, selectProject, addTodoModeSwitch, editTodoModeSwitch})(AddNewProject);