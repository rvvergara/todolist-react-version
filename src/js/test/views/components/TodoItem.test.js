import React from 'react';
import { shallow } from 'enzyme';
import { TodoItem } from '../../../views/components/TodoItem';
import todos from '../../fixtures/todos';

describe('TodoItem', () => {
  test('', () => {

  });
});
