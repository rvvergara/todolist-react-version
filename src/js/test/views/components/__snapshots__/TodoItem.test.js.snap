// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem should render correctly and have correct initial state 1`] = `
<tr
  className=""
  id={11}
>
  <td>
    Second Task
  </td>
  <td>
    Minim proident occaecat nulla ad nisi.
  </td>
  <td>
    Monday May 6, 2019
  </td>
  <td />
  <td>
    Cillum aute eu magna anim id tempor non cupidatat consequat deserunt ad.
  </td>
  <td>
    <input
      checked={false}
      defaultValue={false}
      onChange={[Function]}
      type="checkbox"
    />
  </td>
  <td>
    <button
      className="btn btn-sm btn-warning"
      disabled={false}
      onClick={[Function]}
    >
      Update
    </button>
    <button
      className="btn btn-sm btn-danger"
      disabled={false}
      onClick={[Function]}
    >
      Delete
    </button>
  </td>
</tr>
`;
